### DISCUSSION TITLE
Encoding issues due to missing locale settings

### DISCUSSION THREAD

**Discussion by marcelotto** [Taitava/obsidian-shellcommands#373] [Debugging and testing]
The shell environment (at least with Zsh under MacOS) does not define the `LC_CTYPE` environment variable which is usually set automatically by the system. Also, the `locale LC_CTYPE` returns a very basic `US-ASCII`. This is causing, for example, encoding issues when copying a text with characters outside of this character set to the clipboard with `pbcopy`.

The problem can be seen by creating a shell command `echo $LC_CTYPE ; locale LC_CTYPE` which produces the following output on my system:

```

US-ASCII
``` 

In a normal shell environment, this command produces this output on my system:

```
UTF-8
UTF-8
```

**Comment by Taitava** [comment]
Hi! Thank you for your message! 🙂 

This is valuable information to me, as it makes me learn about how the default shells of this plugin behave. I'm not very familiar with Zsh, and I don't have macOS, so I can't really do much testing.

However, I guess that this happens because when using the normal shells built into the SC plugin, [Node.js's `child_process`](https://nodejs.org/api/child_process.html) does not use a `-l` (login) switch when it invokes a shell. **If** this switch was used, Zsh would process additional configuration files before executing your command. As those configuration files are not processed, I supect it causes the `LC_CTYPE` environment variable not to be defined.

Perhaps you could try to solve this problem by creating a [custom shell configuration](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Custom+shells) for Zsh that includes the `-l` argument:
- [General instructions on configuring custom shells](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Settings+for+custom+shells)
- _Shell name_: For example _Custom Zsh_
- _Executable binary file path_: `/bin/zsh`
- _Shell arguments_:
    ```
    -l
    -c {{!shell_command_content}}
    ```
- _Host operating system_ and _Shell's operating system_: macOS for both.
- _Special characters escaping_: _Unix shell style with \ as escape character_
- _Path translator_: Not needed, so please leave empty.
- _Wrapper for shell command_: Not needed, too.
- After creating the custom shell configuration, please remember to [make your shell commands use the configuration](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Settings+for+custom+shells#Make+shell+commands+use+the+newly+created+shell).

Please let me know if this solved tthe problem or not, or if you have any other questions. 👍

**Reply by marcelotto** [reply]
Sorry for the late reply. I was offline for the past few days due to sickness.

Thanks for pointing out a potential workaround. Unfortunately, this isn't possible in my use case, which is the [Magma](https://github.com/marcelotto/magma) project that creates Obsidian vaults for its users and offers additional commands on top of those vaults (that are triggered with this plugin). In this scenario, however, I actually can not assume a particular shell that users might have.

The workaround I'm currently using is to set the `LC_CTYPE` in a shell wrapper script I'm already using to perform the Magma command calls in a portable manner (https://github.com/marcelotto/magma/commit/c129f4f88edfbe0818df0943a4e5a411ee81bf96).

I just wanted to point out this problem as I can imagine it might affect more users of this plugin.

**Reply by Taitava** [reply]
Oh, that's true, in that case you cannot assume users have a certain shell.

Maybe I should think about changing the builtin shells so that they would use the `-l` (login) switch on Linux and macOS, so that startup scripts get executed. The current state of not running the startup scripts has already caused a lot of problems related to missing directories in `$PATH`, and now this encoding problem appeared.
